enhancements/ingress/forwarded-header-policy.md

```diff
@@ -0,0 +1,339 @@
+	// NeverHTTPHeaderPolicy never sets the header, preserving any existing
+	// header.
+	NeverHTTPHeaderPolicy IngressControllerHTTPHeaderPolicy = "never"
+)
+
+// IngressControllerHTTPForwardedHeaderPolicy describes how the HTTP Forwarded
+// header and related headers are handled.
+type IngressControllerHTTPForwardedHeaderPolicy struct {
+	// policy specifies when and how the IngressController sets the
+	// Forwarded, X-Forwarded-For, X-Forwarded-Host, X-Forwarded-Port,
+	// X-Forwarded-Proto, and X-Forwarded-Proto-Version headers.  The value
+	// may be one of the following:
+	//
+	// * "append", which specifies that the IngressController appends the
+	//   headers, preserving existing headers.
+	//
+	// * "replace", which specifies that the IngressController sets the
+	//   headers, replacing any existing Forwarded or X-Forwarded-* headers.
+	//
+	// * "if-none", which specifies that the IngressController sets the
+	//   headers if they are not already set.
+	//
+	// * "never", which specifies that the IngressController never sets the
+	//   headers, preserving any existing headers.
+	//
+	// By default, the policy is "append".
+	//
+	// +optional
+	Policy IngressControllerHTTPHeaderPolicy `json:"policy,omitempty"`
+}
+```
+
+By default, the IngressController appends `Forwarded` and related headers to
+whatever headers are already present in the HTTP request.  The `Policy` field
+enables the user to modify this behavior, as described in the godoc above.  For
+example, the following IngressController replaces any `Forwarded` and related
+headers before forwarding the request:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpHeaders:
+    forwarded:
+      policy: replace
+```
+
+Routes may override this policy using the
+`haproxy.router.openshift.io/set-forwarded-headers` annotation, which allows the
+same values as the IngressController's `spec.httpHeaders.forwarded.policy`
+field.  For example, an IngressController should never set the `Forwarded`
+header or related headers for the following Route, irrespective of the
+IngressController's own particular policy:
+
+```yaml
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  annotations:
+    haproxy.router.openshift.io/set-forwarded-headers: never
+  name: hello-openshift
+  namespace: hello-openshift
+spec:
+  to:
+    kind: Service
+    name: hello-openshift
+```
+
+#### Validation
+
+Omitting `spec.httpHeaders` or omitting `spec.httpHeaders.forwarded` specifies
+the default behavior.
+
+The API validates the `spec.httpHeaders.forwarded` field value as described by
+the field type's `+kubebuilder:validation:Enum` marker.
+
+The HAProxy configuration template validates the
+`haproxy.router.openshift.io/set-forwarded-headers` annotation.  If the
+annotation specifies an invalid value, then the IngressController uses its
+configured policy.
```

Comment:
> If the annotation specifies an invalid value, then the IngressController uses its configured policy.

Maybe we should reflect the policy used by the ingress controller in status.

Reply:
Do you mean reflect the policy in the ingress controller's status, or in the route's?